src/Dialect/ONNX/Transforms/QlinearConvPass.cpp

```diff
@@ -0,0 +1,327 @@
+
+  biasI32Attrs.reserve(denseBiasF.getNumElements());
+  for (auto apf : denseBiasF.getValues<llvm::APFloat>()) {
+    double f = apf.convertToFloat();
+    double qd = std::nearbyint(f / denom);
+    int64_t qi = static_cast<int64_t>(qd);
+
+    // clamp to int32 range
+    if (qi > std::numeric_limits<int32_t>::max())
+      qi = std::numeric_limits<int32_t>::max();
+    if (qi < std::numeric_limits<int32_t>::min())
+      qi = std::numeric_limits<int32_t>::min();
+
+    biasI32Attrs.push_back(
+        rewriter.getI32IntegerAttr(static_cast<int32_t>(qi)));
+  }
+
+  return biasI32Attrs;
+}
+
+/// Pattern that matches Conv fed by DequantizeLinear(s) and consumed by
+/// QuantizeLinear.
+struct ConvToQLinearConvPattern : public OpRewritePattern<ONNXConvOp> {
+  using OpRewritePattern<ONNXConvOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(
+      ONNXConvOp convOp, PatternRewriter &rewriter) const override {
+    Location loc = convOp.getLoc();
+
+    // Conv input X must come from DequantizeLinear ----
+    auto dqInputOp =
+        dyn_cast_or_null<ONNXDequantizeLinearOp>(convOp.getX().getDefiningOp());
+    if (!dqInputOp) {
+      return failure();
+    }
+
+    Value qInput = dqInputOp.getX();
+    Value xScale = dqInputOp.getXScale();
+    Value xZp = dqInputOp.getXZeroPoint();
+
+    // ---- Check input type ----
+    if (!isTensorI8OrU8(qInput))
+      return failure();
+
+    // Conv weight W must come from DequantizeLinear ----
+    auto dqWeightOp =
+        dyn_cast_or_null<ONNXDequantizeLinearOp>(convOp.getW().getDefiningOp());
+    if (!dqWeightOp) {
+      return failure();
+    }
+
+    Value qWeight = dqWeightOp.getX();
+    Value wScale = dqWeightOp.getXScale();
+    Value wZp = dqWeightOp.getXZeroPoint();
+
+    if (!isTensorI8OrU8(qWeight))
+      return failure();
+
+    // Conv output consumed by QuantizeLinear (qOutOp) ----
+    if (convOp->getUsers().empty() && convOp->getNumResults() != 1)
+      return failure();
+
+    Operation *firstUser = *convOp->getUsers().begin();
+    auto qOutOp = dyn_cast<ONNXQuantizeLinearOp>(firstUser);
+    if (!qOutOp)
+      return failure();
+    Value yScale = qOutOp.getYScale();
+    Value yZp = qOutOp.getYZeroPoint();
+
+    if (!isScalarF32(xScale) || !isScalarF32(wScale) || !isScalarF32(yScale)) {
+      return failure();
+    }
+
+    if (!isScalarI8OrU8(xZp) || !isScalarI8OrU8(wZp) || !isScalarI8OrU8(yZp)) {
+      return failure();
+    }
+
+    Value biasVal = convOp.getB();
+
+    Value biasInt32Val;
+    if (!biasVal || isa<ONNXNoneOp>(biasVal.getDefiningOp())) {
+      // Case 0: No bias at all
+      biasInt32Val = Value();
+    } else {
+
+      // Conv has bias B and its defining op is DequantizeLinear
+      auto dqBiasOp = dyn_cast<ONNXDequantizeLinearOp>(biasVal.getDefiningOp());
+      if (!dqBiasOp) {
+        return failure();
+      }
+
+      Value biasQ = dqBiasOp.getX();
+      auto biasQType = biasQ.getType().dyn_cast<mlir::RankedTensorType>();
```

Comment:
Please do not use the `cast` member funtions they are deprecated (and cause compiler warnings)
```suggestion
      auto biasQType = dyn_cast<mlir::RankedTensorType>(biasQ.getType());
```